### ISSUE TITLE
[tune] Cross Validation (simply parellization) patterns? 

### ISSUE THREAD

**Issue by cottrell** [ray-project/ray#7744]
Something similar was asked before but this is different. 

What pattern is imagined for run dags that have one config instance leading to a number of (model, data_train, data_test) runs? You might typically collapse this into a mean score - std score or something like that. 

I feel like I am fighting the framework so am probably not getting something right. The issue is the Trainable class handle the serialization/persistance but you really way to pass everything down to the remotes that get parallized.

Or is there some pattern at the tune.run level that allows you to sample across the train/test pairs (not optimize) i.e. treat the data like config params?

https://github.com/ray-project/ray/issues/6560

**Comment by anabranch** [comment]
Hey David, great question. Would you be open to a quick call on this subject - it's something we've been discussing internally (the ray / anyscale team) and it'd be good to understand your perspective to make sure we're thinking about it in the right way - given your well formulated question :).

can you ping me, bill @ anyscale or if you're on the ray slack I'm on there too.

**Comment by richardliaw** [comment]
Hey @cottrell, would this work https://ray.readthedocs.io/en/latest/tune-searchalg.html#repeated-evaluations for you? 

Documentation here: https://ray.readthedocs.io/en/latest/tune/api_docs/suggestion.html#ray.tune.suggest.Repeater

What you could do is you can have the Trainable execute something different depending on the trial_index. 

```python
class TestMe(Trainable):
    def _setup(self, config):
        index = config[tune.suggest.repeater.TRIAL_INDEX]
        data_train, data_test = create_from_index(index, config)

    def _train(self):
        ...

tune.run(TestMe, search_alg=Repeater(HyperOptSearch(search_space), repeat=5, set_index=True))
```

Does this make sense? Feel free to follow up with any questions (or any suggestions for how we can improve the docs).

**Comment by cottrell** [comment]
@richardliaw I think a modified Repeater would handle that case I'm thinking of ... but I'm kind of reluctant to fit a framework around it. You could get really fancy and just treat all individual runs as independent and just update the stats scoring mechanism in the scheduler I guess. Will try to jump on slack

**Comment by cottrell** [comment]
Can do a call to chat. @anabranch Will msg you @ anyscale ... I've requested to join slack but it might take a few days.

**Comment by richardliaw** [comment]
Sent an invite; happy to chat online.

**Comment by richardliaw** [comment]
I think we resolved this offline (feel free to reopen if not resolved.)

**Comment by antonwnk** [comment]
Curious what the specific resolution of this was. @cottrell, have you settled for the Repeater solution Richard showed above or have you found a different way around it?

**Comment by richardliaw** [comment]
Ah, I think the resolution was to just not use this type of computation pattern.

**Comment by bbalaji-ucsd** [comment]
I'm interested in this issue. Is there an example of doing cross validation with tune? I'm interested in doing cross validation with tensorflow or torch frameworks.

**Comment by FarzanT** [comment]
I'm also interested! @richardliaw, is the `Repeater` class the best way of performing cross-validation with Ray Tune? Thanks!

**Comment by FarzanT** [comment]
@richardliaw Hi, do you have any advice on using the Repeater for cross-validation? If not, do you have another technique in mind? 
Thank you!

**Comment by richardliaw** [comment]
Hey @FarzanT sorry for the slow reply! Repeater would allow you to run the same model/hypers but using an index to do a different split.

In general I would suggest just doing the cross validation manually within the training run. This is what we do in `tune-sklearn`: https://github.com/ray-project/tune-sklearn/blob/master/tune_sklearn/_trainable.py#L176

**Comment by krfricke** [comment]
Here's a concept that might work:

https://discuss.ray.io/t/ray-tune-confidence-interval/2967

Let me pull the code from the example:

```
from ray import tune
import numpy as np

def train(config):
    print(config["repeat"], config["a"])
    return np.random.randn() * 0.1 + config["a"]


searcher = tune.suggest.basic_variant.BasicVariantGenerator(
    constant_grid_search=True
)

analysis = tune.run(
    train,
    config={
        "repeat": tune.grid_search(list(range(4))),
        "a": tune.uniform(0, 1)
    },
    search_alg=searcher,
    num_samples=2)

df = analysis.dataframe()

print(df.groupby(["config/a"]).mean())
```

Basically what you could do is to use the `repeat` config parameter to select a different subset of the data to perform cross-validation on. By fetching the resulting dataframes you can then calculate metrics like mean, std, etc.

This was enabled by the `constant_grid_search` parameter introduced in https://github.com/ray-project/ray/pull/16501.

I'll close this issue for now, but if there's any more questions or suggestions around this topic, please feel free to re-open or create a new discuss/issue.

**Comment by mlguruz** [comment]
> Hey @FarzanT sorry for the slow reply! Repeater would allow you to run the same model/hypers but using an index to do a different split.
> 
> In general I would suggest just doing the cross validation manually within the training run. This is what we do in `tune-sklearn`: https://github.com/ray-project/tune-sklearn/blob/master/tune_sklearn/_trainable.py#L176

Hi, if we do it *manually* then there isn't any parallelization across CV folds, right? 

Is there any way to run different folds **in parallel** *within* a Trial? 

Or, I guess a more general question is, can Tune allow us to have a concept of Trial dependencies? 

Appreciate if you could share your thoughts on this!

**Comment by zcarrico-fn** [comment]
@krfricke , thank you for the example! How would you recommend this be done if the `Searcher` is `BayesOptSearch`, which will raise an exception if tune.grid_search is passed? (Although the `Repeater` [strategy](https://github.com/ray-project/ray/issues/7744#issuecomment-604080073) will technically work with `BayesOptSearch` it's not appropriate because the prior would become the metric average from training on different data splits, rather than the same data split)
